### ISSUE TITLE
Feedback

### ISSUE THREAD

**Issue by jenshaase** [mabasic/phel-json#1]
Here is some feedback:

* Do you also plan to a `decode` method?
* Keywords should not be encoded without a `:` before. You can use `(php/-> keyword (getName))` to get the name of the keyword without the `:`
* Symbols should be converted to a string. You can use  `(php/-> symbol (getName))`
* I would recommend the `for` loop over the `foreach` loop. Even better would be `reduce`. `(reduce (fn [arr [k v]] (do (php/aset arr k v) arr)) (php/array) @{"a" 1 "b" 2})`

**Comment by mabasic** [comment]
Nice tips! I'll get right to it.

Yes, the decode method is in the plan.

**Comment by mabasic** [comment]
@jenshaase Can you give me some examples on what is a symbol?

If I pass a symbol (variable or function) it gets resolved before being encoded.

**Comment by jenshaase** [comment]
In the repl/code you can do this:

`'s` is a symbol
`(symbol? 's)` is true

**Comment by mabasic** [comment]
@jenshaase Why is reduce better than foreach? 

Also, the documentation for `reduce` is very limited `(reduce f init xs)`. Reduce usually means to reduce an array to one value, like a sum or accumulator...

I am getting this error when using reduce:

```
Test: (= [1,2,3] (encode [1 2 3]))
 throwed exception: InvalidArgumentException

InvalidArgumentException: cannot call 'next on integer
```

Easy way to recreate:

```
>>> (reduce (fn [arr [k v]] (do (php/aset arr k v) arr)) (php/array) [1 2 3])
InvalidArgumentException: cannot call 'next on integer
```

Reduce does not work on tuples or sets?

I've pushed new code, so take a look.

**Comment by jenshaase** [comment]
Reduce is not better that the others. It was just an idea.

On arrays/tuples and sets reduce works like this:
```
(reduce (fn [arr v] (do (php/aset arr k v) arr)) (php/array) [1 2 3])
```
The second element is not a list of key value pairs but just the value. The problem for you is probably that you don't get the index of the array within the reduce function.

**Comment by mabasic** [comment]
@jenshaase I have tried with reduce, but since I can't get the index (key) it is useless for my use case.

I have also tried with `for`, but it always returns a Phel array and inside it another datastructure, which really complicates things. I would have to use another `for` to deconstructure the generated array...

The solution with `foreach` seems simplest and uses less lines of code.

See [file](https://github.com/mabasic/phel-json/blob/master/src/json.phel) for the things that I have tried.

I am ready to tackle `decode` function now.

**Comment by jenshaase** [comment]
Ok makes sense. One other optimization. I think you don't need `datastructure?`. Instead I would recommend to use the php function `is_iterable` (`(php/is_iterable x)`).

The code the would also work for PHP `ArrayIterator`, `Generator` and `ArrayObject`.

**Comment by mabasic** [comment]
Implemented. It is better now. I've added a test for when the Exception is thrown.

My goal for the encode function is to accept the same parameters as PHP `json_encode`: https://www.php.net/manual/en/function.json-encode.php

Is there a way to set the function parameter to have a default value in Phel?

```
json_encode ( mixed $value , int $flags = 0 , int $depth = 512 )
```

**Comment by jenshaase** [comment]
```
(if (php/is_iterable k) (throw (php/new Exception "Key cannot be iterable.")))
```
This doesn't catch all errors. What if `k` is `stdclass` or `\DateTime` for example. I would check for values that are allowed as key and not for values that are not allowed.

Regarding the parameter. I would do it like this:
```
(defn encode [value & @{:flags flags :depth depth}]
  (let [flags (or flags 0)
         depth (or depth 512)]
     ...))
```

Then you can call it like this:
```
(encode my-value {:flags 1 :depth 100})
(encode my-value {:depth 100})
(encode my-value {:flags 1})
(encode my-value)
```

**Comment by mabasic** [comment]
@jenshaase I've just pushed a new commit. Can you take a look now?

I have improved tests and added Exceptions for different things.

**Comment by jenshaase** [comment]
Some improvements:

```
(if (not (or (int? k) (float? k) (symbol? k) (keyword? k) (string? k)))
```
I would move this to a function `valid-key?`
I would recommend to use `when` instead of `if` if you have handle just one case.
Instead of `(when (not (valid-key? k)))` you could also use `(when-not (valid-key? k))`. This is much more readable.

Maybe rename `phel-to-php` to `encode-value`?

For all the if statements in `encode` I also would suggest `when`.

```
# TODO: Find out what does depth do in json_encode?
```
This is the maximum nested array depth. For example
```
[
  [
    [1]
  ]
]
```
has depth 3.

**Comment by mabasic** [comment]
@jenshaase All nice tips! I feel like I'm learning much more about the language this way. Did not know that `when` and `when-not` even existed.

I have a question about depth. I am still unclear about the expected output.

If I have a nested array 3 levels deep as in your example, what should be the JSON if depth is set to `2`. When I run tests, it returns `false`. What is the purpose of depth then? It will return JSON only if the depth of the given array is lower then the depth specified? That does not make sense to me. What is the use case for something like that?

**Comment by mabasic** [comment]
@jenshaase I've added `(json/decode)` function. Not the library can encode from phel datastructures to JSON and decode JSON to phel datastructure.

I need to add a bit more tests, but it seems to be working. What do you think?

**Comment by jenshaase** [comment]
```
(put table (keyword k) (decode-value v))
```
I wouldn't convert `k` to a `keyword`. `k` could not only be a string but also a number.

**Comment by mabasic** [comment]
> ```
> (put table (keyword k) (decode-value v))
> ```
> I wouldn't convert `k` to a `keyword`. `k` could not only be a string but also a number. 

It can't be a number. It can be string and a number inside string. That is still string.

If keyword function can't convert string "3" to keyword then I I will need to change the code so that the keys are strings instead of keywords.

But I would prefer them to be keywords...

**Comment by jenshaase** [comment]
It can be a number. Not because of JSON but because of PHP ;)
```php
$x = json_decode('{"123": "test"}', true);
var_dump(array_keys($x)); 
/*
array(1) {
  [0] =>
  int(123)
}
*/
```
But if you prefer them to be keyword it is ok for me. Code should still work.

**Comment by mabasic** [comment]
I need to test this, but I think that array_keys is converting string "123" to integer in your example. JSON keys must be strings. If PHP json_decode converts them to integers that should be wrong.

Why would json_decode convert keys from strings to integers...

**Comment by mabasic** [comment]
Just released version `0.1.0` based on Phel 0.2.0. Going to close this issue for now. Thank you for all the feedback.
